File: 2015/20/python/main.py
from pathlib import Path

def min_houses(target: int) -> int:
    max_house = target // 10
    presents = [0] * (max_house + 1)
    
    for elf in range(1, max_house + 1):
        for house in range(elf, max_house + 1, elf):
            presents[house] += elf * 10
    for house, total in enumerate(presents):
        if total >= target:
            return house
    return -1

def min_11_houses(target: int) -> int:
    max_house = target // 11
    presents = [0] * (max_house + 1)
    
    for elf in range(1, max_house + 1):
        count = 0
        for house in range(elf, max_house + 1, elf):
            presents[house] += elf * 11
            count += 1
            if count == 50:
                break
    for house, total in enumerate(presents):
        if total >= target:
            return house
    return -1

def main() -> None:
    data = int(Path("../input.txt").read_text().strip())
    
    print("=" * 20, "Part 1", "=" * 20)
    house = min_houses(data)
    print(f"The first house to receive at least {data} presents is house {house}")
    
    print("=" * 20, "Part 2", "=" * 20)
    house = min_11_houses(data)
    print(f"The first house to receive at least {data} presents is house {house}")

if __name__ == "__main__":
    main()

File: 2015/20/rust/src/main.rs
use std::fs;


fn min_houses(target: &i32) -> i32 {
    let max_house = target / 10;
    let mut presents = vec![0; (max_house + 1) as usize];

    for elf in 1..max_house + 1{
        for house in (elf..=max_house).step_by(elf as usize){
            presents[house as usize] += elf * 10;
        }
    }
    for (house, total) in presents.iter().enumerate(){
        if total >= &target{
            return house as i32;
        }
    }
    -1
}

fn min_11_houses(target: &i32) -> i32 {
    let max_house = target / 10;
    let mut presents = vec![0; (max_house + 1) as usize];

    for elf in 1..max_house + 1{
        let mut count = 0;
        for house in (elf..=max_house).step_by(elf as usize){
            presents[house as usize] += elf * 11;
            count += 1;
            if count == 50{
                break
            }
        }
    }
    for (house, total) in presents.iter().enumerate(){
        if total >= &target{
            return house as i32;
        }
    }
    -1
}

fn main() {
    let data = fs::read_to_string("../input.txt")
    .expect("Error reading input file")
    .trim()
    .parse::<i32>()
    .expect("Failed to parse input to i32");

    let sep = "=".repeat(20);
    let part_1 = min_houses(&data);
    let part_2 = min_11_houses(&data);
    println!("{} Part 1 {}\nThe first house to recive at least {} presents is house nº {}", sep, sep, data, part_1);
    println!("{} Part 2 {}\nThe first house to recive at least {} presents is house nº {}", sep, sep, data, part_2);
}
